# BRAPeS

We present BRAPeS (BCR Reconstruction Algorithm for Paired-End Single-cell), a software for reconstruction of B cell receptors (BCR) using short, paired-end single-cell RNA-sequencing. <br />  

BRAPeS is an extension of [TRAPeS](https://github.com/YosefLab/TRAPeS), and reconstruct the BCR in 5 steps: For each chain, it first identify the V and J segments by searching for paired reads with one read mapping to the V segment and its mate mapping to the J segment. Then, a set of putative CDR3-originating reads are identified as the set of unmapped reads whose mates map to the V,J and C segments. Next, an iterative dynamic programming algorithm is used to reconstruct the CDR3 region with the putative CDR3 reads. The isotype of the BCR is then determined by running RSEM on the reconstructed sequence with all possible constant segments added to it. Finally, BRAPeS corrects for somatic hypermutations by collecting all reads aligning to the genomic CDR1 and CDR2 sequences and aligning the reads against themselves to obtain a reconstruction of the consensus sequence. <br />  

For more information, see our paper on [bioRxiv](https://www.biorxiv.org/content/early/2018/08/10/389999)
<br />  

##installing BRAPeS <br />  

BRAPeS is written in Python and C++, and works on Linux. BRAPeS require the following python libraries: <br />  

-	biopython  <br />
-	pysam  <br />

In addition, BRAPeS requires the following software:  <br />

-	bowtie2  <br />
-	RSEM  <br />
- samtools <br />

### before running BRAPeS  <br />
BRAPeS takes as input mapped and unmapped files of reads after genomic alignment (e.g. using TopHat).<br />
BRAPeS assumes a certain folder structure: It assumes that each cell has its own folder, and all of those folders are under one path. Also, it assumes that each cell folder has identical subfolder structure.<br />
In case the genomic alignment was done with a software that output only one bam file (e.g. STAR), you can provide BRAPeS with the same file as both the mapped and unmapped file. Please email us if this produces an error.<br />
 
##Running BRAPeS <br />

To run BRAPeS, simply run:<br />
 
>python  brapes.py \[options\] <br />

To display help: <br />

>python  brapes.py -h <br />
 
###Options when running BRAPeS <br />

**Input files:** <br />
<br />

-path : The folder with all the single cell samples. Assumes every subfolder under this path is a folder of a single cell sequencing results. <br />
-bam : The location of the sorted mapped bam file relative to the single cell folder. For example, if under each single cell folder there is a subfolder named “TopHat\_Output” and the mapped file is under that folder and named mapped.bam, the command should read “–bam TopHat\_Output/mapped.bam”. <br />
-unmapped : the location of the bam file containing the unmapped reads. Similar to the –bam tag, it is relative to the single cell folder. <br />
<br />
**Output files:** <br />
<br />
-output : Output prefix for the files generated by BRAPeS in every single cell folder, relative to the single cell folder (e.g. TopHat_Output/BRAPeS/TCR.out). Can include a subdirectory (BRAPeS will create that subdirectory if it does not yet exists). <br />

-sumF : Prefix for the output summary files of the entire path (summary of all single cells together). <br />
<br />
**Parameters for the BCR (CDR3) reconstruction** <br />
<br />
-downsample : For very deep libraries, there can be many BCR-originating reads which can increase running time significantly. By adding this tag, in case there are more than 5,000 reads mapping to the V and J segment, and more than 5,000 putative CDR3-originating reads, BRAPeS will downsample the reads to include only 5,000 V-J reads and only 5,000 CDR3-originating reads for the reconstruction process. This will not influence the read mapping with RSEM to find the isotype. However, to further decrease running time the number of aligned reads reported at the summary file are underestimated. <br />

-score : The threshold score for the alignment of a read to the V or J segment. Any putative CDR3-originating read that its alignment to the V or J segments passes this score will be used for reconstruction. Default is 15, but should be changed based on sequencing quality and length. <br />

-overlap : Minimum number of bases that the extended V and J segments should overlap in order to stop the reconstruction. Default is 10. <br />

-iterations : Maximum number of times to extend the V and J segments. If after that number of iterations the V and J segments still do not overlap the reconstruction is determined as unsuccessful. Default is 20. <br />  

-bases : Number of bases from the V and the J segments that will be used as the initial template for the reconstruction. Default is min(length(V), length(J)). <br />

-NoLowQ	: By default, the putative CDR3-originating reads are identified as unmapped reads whose mate is aligned to the V/J/C segments. However, those reads can by chance be mapped to other places in the genome (usually low quality mapping). By including the –NoLowQ tag the set of CDR3-origintating reads will exclude the reads that map to other places in the genome. <br />

-top: Rank all the V-J pairs based on the number of mapped reads, and reconstruct only the top X number of V-J pairs. Default: reconstruct all V-J pairs. Recommended for very deep libraries or libraries when there are many possible V-J pairing. <br />

-byExp: Must be used along with the top parameter. Rank all the V-J pairs based on the number of mapped reads, but instead of reconstructing only the top X number of V-J pairs, randomly choose 2 V-J pairs with the same rank (same number of mapped reads) and reconstruct them. Then BRAPeS will move on to the next rank until the number of V-J pairs specified with top has been reconstructed. Default: off. <br />

-oneSide: Add this parameter to also search for productive reconstructions only from the extended V segment (in case of no overlap between the extended V and extended J segments). Default: off. <br />

<br />

**Parameters for the CDR1 and CDR2 reconstruction** <br />
<br />
Please note the BRAPeS currently reconstruct only the heavy chain hypervariable region. An updated version with light chain CDR1/2 reconstruction will be uploaded very soon.
<br />
<br />
-skipHVR: Add this flag to avoid reconstructing the hypervariable regions. <br /> 
-HVR_extension: The number of bases to extend the CDR1 and CDR2 from both sides for alignment. Default is 15. <br />
-HVR_score: The minimum alignment score of  a read to the CDR1/CDR2 in order for it to be used for reconstruction. Default is 15. <br />
-HVR_min_reads: The minimum number of reads that needs to align to the CDR1/2 in order to perform CDR1/2 reconstruction. If this number is not met the genomic sequence will be returned. Default is 10. <br />
-HVR_max_reads: The maximum number of reads that needs to align to the CDR1/2. Once this number is met the algorithm will move on to reconstruct the CDR1/2 sequence. Default is 200. <br />
-HVR_moveOn: The maximum number of reads that BRAPeS will randomly sample for the set of putative reads and attempt to align to the CDR1/2. Default is 30,000. <br />
<br />

**Paths to other software:** <br />
<br />
-bowtie2 : Path to bowtie2. If not used assumes bowtie2 is in the default PATH <br />

-rsem: Path to RSEM. If not used assumes RSEM is in the default PATH <br />

-samtools: Path to samtools. If not used assumes samtools is in the default PATH <br />
<br />
**Other parameters:** <br />
<br />

-genome : The genome used for genomic alignment. By default, BRAPeS only supports hg38 and mm10, and for mm10 with NCBI chromosome naming use mm10_ncbi.. However, BRAPeS can be used for any genome or any organism, as long as the genomic annotations for the V/J/C segments are available. In order to run BRAPeS on genomes besides mm10 or hg38, the user must create the relevant annotation files [(see here for more information)](https://github.com/YosefLab/BRAPeS/tree/master/Data). In addition, this will require to add the '-Hminus'/'-Kminus'/'-Lminus' parameters if necessary. <br />

-Hminus: Only used for user supplied genomes. Add this flag if the (majority) of V and J annotations of the heavy chain are on the minus strand. <br />
-Kminus: Only used for user supplied genomes. Add this flag if the (majority) of V and J annotations of the kappa chain are on the minus strand. <br />
-Lminus: Only used for user supplied genomes. Add this flag if the (majority) of V and J annotations of the lambda chain are on the minus strand. <br />

-strand : Strand orientation of the reads, options are [minus, plus, none]. For transcripts on the positive strand, to which strand does the rightmost (in genomic coordinates) mate of the read map to. Default is minus. <br />
<br /><br />

## BRAPeS output<br />
<br />
- sumF.summary.txt : Summary of the reconstruction status in each cell (successful/unsuccessful). <br />

- sumF.BCRs.txt : A list of all reconstructions (productive, unproductive and partial) of all cells. <br />
<br />
In addition, in each single cell folder you can find the following output files: <br />
<br />
-	output.\[heavy/kappa/lambda\].junctions.txt : The set of V-J pairs found (before reconstruction) <br />
-	output.reconstructed.junctions.\[heavy/kappa/lambda\].fa : the set of reconstructed junction. If reconstruction was unsuccessful, the partial V and J reconstructions will be separated by N’s. <br />
-	output.\[heavy/kappa/lambda\].mapped.and.unammed.fa : the set of the putative CDR3-originating reads used for the reconstruction.  <br />
-	output.\[heavy/kappa/lambda\].\[R1/R2\].fa : Set of paired-end reads that are aligned to the reconstructed BCRs in order to quantify the expression of each BCR using RSEM. <br />
-	output.\[heavy/kappa/lambda\].rsem.out* : The output files created by RSEM. <br />
-	output.\[heavy/kappa/lambda\].full.BCRs.fa : Fasta file with the full BCR sequences. Includes all possible isotypes <br />
-	output.\[heavy/kappa/lambda\].full.BCRs.bestIso.fa : Fasta file with the full sequences of the BCRs, after choosing only the highly expressed isotype. <br />
- output.\[heavy/kappa/lambda\].full.BCRs.bestIso.CDR1.CDR2.reconstructions.fasta : The full BCR sequences after CDR1 and CDR2 reconstruction. <br />
- output.\[heavy/kappa/lambda\].full.BCRs.bestIso.hypervariable.regions.fasta : The CDR1 and CDR2 reconstructed sequences. <br />
-	output.summary.txt : Summary of all the reconstructed chains in this cell. <br />

<br /><br />

For any questions or comments, please email Shaked Afik, safik@berkeley.edu.
